addons/fleet/fleet.py

```diff
@@ -1,818 +1,725 @@
+        from the start_date to today
+        The created costs are associated to a contract thanks to the many2one field contract_id
+        If the contract has no start_date, no cost will be created, even if the contract has recurring costs
+        """
+        VehicleCost = self.env['fleet.vehicle.cost']
+        today = fields.Datetime.from_string(fields.Date.today())
+        log_contract = self.env['fleet.vehicle.log.contract'].search([('state', '!=', 'closed'),
+                                                                      '|', ('start_date', '=', None),
+                                                                      ('cost_frequency', '!=', 'no')])
+        deltas = {'yearly': relativedelta(years=+1), 'monthly': relativedelta(months=+1),
+                  'weekly': relativedelta(weeks=+1), 'daily': relativedelta(days=+1)}
+        for contract in log_contract:
             last_cost_date = contract.start_date
             if contract.generated_cost_ids:
-                last_autogenerated_cost_id = vehicle_cost_obj.search(cr, uid, ['&', ('contract_id','=',contract.id), ('auto_generated','=',True)], offset=0, limit=1, order='date desc',context=context, count=False)
-                if last_autogenerated_cost_id:
-                    found = True
-                    last_cost_date = vehicle_cost_obj.browse(cr, uid, last_autogenerated_cost_id[0], context=context).date
-            startdate = datetime.datetime.strptime(last_cost_date, tools.DEFAULT_SERVER_DATE_FORMAT).date()
-            if found:
-                startdate += deltas.get(contract.cost_frequency)
-            while (startdate <= d) & (startdate <= datetime.datetime.strptime(contract.expiration_date, tools.DEFAULT_SERVER_DATE_FORMAT).date()):
+                vehicle_cost = VehicleCost.search(['&', ('contract_id', '=', contract.id),
+                                                 ('auto_generated', '=', True)], limit=1, order='date desc')[0]
+                if vehicle_cost:
+                    last_cost_date = vehicle_cost.date
+            last_cost_date = datetime.datetime.strptime(last_cost_date, tools.DEFAULT_SERVER_DATE_FORMAT)
+            last_cost_date += deltas.get(contract.cost_frequency)
+            while (last_cost_date <= today) and (
+                    last_cost_date <= fields.Datetime.from_string(contract.expiration_date)):
                 data = {
                     'amount': contract.cost_generated,
-                    'date': startdate.strftime(tools.DEFAULT_SERVER_DATE_FORMAT),
+                    'date': last_cost_date.strftime(tools.DEFAULT_SERVER_DATE_FORMAT),
                     'vehicle_id': contract.vehicle_id.id,
                     'cost_subtype_id': contract.cost_subtype_id.id,
                     'contract_id': contract.id,
                     'auto_generated': True
                 }
-                cost_i
```

Comment:
Please check whether this is necessary.
`sum_cost` is a computed field, onchanges are automatic on computed fields.